### PR TITLE
Show custom label in quick open

### LINKED ISSUE #209662: Show Custom Labels in Go to file panel (Ctrl+P)
The new custom labels in the editor tabs is great but would it also be possible to show these custom labels when searching for files in the Ctrl+P menu? I am frequently searching for files in a codebase that has many duplicated file names and it would help find the specific one I am looking for.

I'd be happy to look at opening a pr for this if there is interest.

Thanks for your time.

### PR DESCRIPTION
**Pull request by jswillard**
Fixes https://github.com/microsoft/vscode/issues/209662

Shows the custom labels as defined by `workbench.editor.customLabels.patterns` in the quick open menu. If no pattern matches or custom labels is disabled, then it shows the original name for the URI (filename.extension).

### CODE REVIEW

**Comment by TylerLeonhardt** [comment]
Can you attach a screenshot?

**Comment by jswillard** [comment]
Screenshot 2024-04-05 at 14 25 52
Screenshot 2024-04-05 at 14 27 15


with `workbench.editor.customLabels.patterns` set to this:
```
"workbench.editor.customLabels.patterns": {
    "src/vs/workbench/services/**": "${filename}.${extname} (workbench/services)"
}
```

**Comment by benibenj** [comment]
Currently, the only consumer of `customEditorLabelService` is the editor input. As the editor input already had caching and is able to do it much better then the service could, we did not add caching to `customEditorLabelService`. 

In general the performance of custom labels is not very bad as we only support glob patterns, however, with the amount of lookups/renderings that quickpick does, it might be beneficial to add MRU caching to the service. (Quick pick will probably still have high cache miss rate as it looks up many different files once).

@jswillard if you are up to the challenge you can give this a try. I think we should have some type of support for this in `map.ts`, but if not, I'll look into this.

**Comment by jswillard** [comment]
@benibenj I added a new `MRUCache` type to `map.ts` building off the existing `LRUCache`. I wasn't sure what size of cache to use for this service so I picked 1000, you may want to fiddle with that value. Hopefully this addresses your performance concerns.

File: src/vs/workbench/services/editor/common/customEditorLabelService.ts [hunk 1]
```diff
@@ -115,7 +119,18 @@ export class CustomEditorLabelService extends Disposable implements ICustomEdito
...
 		}
-		return this.applyPatterns(resource);
+
+		const cached = this.cache.get(resource);
+		if (cached !== undefined) {
+			return cached;
+		}
+
+		const result = this.applyPatterns(resource);
+		if (result !== undefined) {
```
Comment (benibenj) [inline src/vs/workbench/services/editor/common/customEditorLabelService.ts:129]: I would add it to the cache even when the result of the computation is undefined, so the next time it checks the file path it knows there is no matching pattern and does not have to do any computations

File: src/vs/workbench/services/editor/common/customEditorLabelService.ts [hunk 2]
```diff
@@ -115,7 +119,18 @@ export class CustomEditorLabelService extends Disposable implements ICustomEdito
 		if (!this.enabled) {
```
Comment (benibenj) [inline src/vs/workbench/services/editor/common/customEditorLabelService.ts:119]: Let's not fill the cache if the user is not using this feature (when he doesn't have any patterns). You can add `this.patterns.length === 0` to the `!this.enabled` check and remove the pattern length check in `applyPatterns()`

File: src/vs/workbench/services/editor/common/customEditorLabelService.ts [hunk 3]
```diff
@@ -39,6 +40,8 @@ export class CustomEditorLabelService extends Disposable implements ICustomEdito
 	private patterns: ICustomEditorLabelPattern[] = [];
 	private enabled = true;
 
+	private cache = new MRUCache<URI, string>(1000);
```
Comment (benibenj) [inline src/vs/workbench/services/editor/common/customEditorLabelService.ts:43]: Even though the URI reference would be more performant as the key, I think we should use the unique identifier of an URI as it is possible the `getName()` method will be called from different places for the same file. This means the key would be a `string`. 
The `URI` class has a `toString()` method which creates a unique identifier of the resource, use this method to get the key.

**Review by benibenj — CHANGES_REQUESTED** [review]
Thanks for the changes! It looks very good. There are just some minor changes required, as soon as these are addressed we can merge this PR :)
Setting the size to 1000 sounds fine for me for the start.

File: src/vs/base/common/map.ts [hunk 4]
```diff
@@ -640,15 +663,53 @@ export class LRUCache<K, V> extends LinkedMap<K, V> {
...
+	}
+
+	override set(key: K, value: V): this {
+		super.set(key, value);
+		this.checkTrim();
+		return this;
+	}
+}
+
+export class MRUCache<K, V> extends Cache<K, V> {
```
Comment (bpasero) [inline src/vs/base/common/map.ts:695]: Can we have a test in `map.test.ts`

File: src/vs/workbench/contrib/search/browser/anythingQuickAccess.ts [hunk 5]
```diff
@@ -952,7 +954,7 @@ export class AnythingQuickAccessProvider extends PickerQuickAccessProvider<IAnyt
 			icon = resourceOrEditor.getIcon();
 		} else {
 			resource = URI.isUri(resourceOrEditor) ? resourceOrEditor : resourceOrEditor.resource;
-			label = basenameOrAuthority(resource);
+			label = this.customEditorLabelService.getName(resource) || basenameOrAuthority(resource);
```
Comment (bpasero) [inline src/vs/workbench/contrib/search/browser/anythingQuickAccess.ts:957]: I am not entirely sure this is a good location to compute the custom editor label: this method will be called not only for recently opened editors, but also all search results. This leads to quite some inconsistency:
* you get the custom label for file results but you cannot search for it
* you see custom labels for file resources but only here, not in any other part of the UI (e.g. explorer)

I would suggest to limit the use of custom editor labels to entries in the picker that are "recently opened" editors.

Screenshot 2024-04-07 at 19 43 32
Screenshot 2024-04-07 at 19 43 07

**Review by bpasero — DISMISSED** [review]
Sorry for late review, I am back from vacation.

File: src/vs/workbench/contrib/search/browser/anythingQuickAccess.ts (hunk 5 above)
Comment (benibenj) [reply src/vs/workbench/contrib/search/browser/anythingQuickAccess.ts:957]: I think it makes more sense to show it for all and not only `recently opened`. We will need to figure out where it makes sense to add the custom labels and where not, but for me it makes sense to have it here. Being able to search for them makes sense to me, but I don't think we can allow it from a performance perspective (we could try and measure but let's not do that now)

Comment (benibenj) [reply src/vs/workbench/contrib/search/browser/anythingQuickAccess.ts:957]: We could consider appending the file name to the description when there is a custom label present. This would allow the user to see both the real and custom names

**Review by benibenj — DISMISSED** [review]

**Comment by benibenj** [comment]
@TylerLeonhardt it seems like the label is the value which is used for searching/filtering the files. 
For example: I have a repo with many `index.jsx` files. I renamed them to `${dirname}.jsx`. When I now search for `index.jsx` in the file picker, it does not show any results anymore. What do you think should be the expected behaviour? Ideally it would be possible to search for both the original and custom name.

File: src/vs/workbench/contrib/search/browser/anythingQuickAccess.ts [hunk 6]
```diff
@@ -952,8 +954,9 @@ export class AnythingQuickAccessProvider extends PickerQuickAccessProvider<IAnyt
 			icon = resourceOrEditor.getIcon();
 		} else {
 			resource = URI.isUri(resourceOrEditor) ? resourceOrEditor : resourceOrEditor.resource;
-			label = basenameOrAuthority(resource);
-			description = this.labelService.getUriLabel(dirname(resource), { relative: true });
+			const customLabel = this.customEditorLabelService.getName(resource);
+			label = customLabel || basenameOrAuthority(resource);
+			description = this.labelService.getUriLabel(!!customLabel ? resource : dirname(resource), { relative: true });
```
Comment (benibenj) [inline src/vs/workbench/contrib/search/browser/anythingQuickAccess.ts:959]: @TylerLeonhardt I propose including the original file name in the description when the label is custom. This allows searching for both original and custom name/label.

**Review by benibenj — DISMISSED** [review]

File: src/vs/workbench/common/editor/resourceEditorInput.ts [hunk 7]
```diff
@@ -116,10 +120,14 @@ export abstract class AbstractResourceEditorInput extends EditorInput implements
 		}
 	}
 
+	private get descriptionResource(): URI {
```
Comment (bpasero) [inline src/vs/workbench/common/editor/resourceEditorInput.ts:123]: Whats the idea here, why this change?

Comment (benibenj) [reply src/vs/workbench/common/editor/resourceEditorInput.ts:123]: We very often show the description (as hover or label description) of the editor input when we show the name of the editor input. If the name is using a custom label then we should show the original file name in the description. 
When there is no custom label, then it's easy to know which file is being represented, it's just `description/name`, but when there is a custom label, this is not the case anymore, so we should make sure the description contains all the information.

Comment (bpasero) [reply src/vs/workbench/common/editor/resourceEditorInput.ts:123]: @benibenj I think I would limit this to hover only, thus `getTitle` but not description. Otherwise, wouldn't the description often repeat what the name already shows?

Comment (benibenj) [reply src/vs/workbench/common/editor/resourceEditorInput.ts:123]: there is a good chance it will repeat, but it won't always be the case. Some users have `${dirname}.${extname}` for example. I'll remove it for the time being and check what users say.

Comment (bpasero) [reply src/vs/workbench/common/editor/resourceEditorInput.ts]: Thanks!

**Review by benibenj — APPROVED** [review]

**Comment by jswillard** [comment]
@TylerLeonhardt is there anything else that needs to be done for this to be merged?

**Review by TylerLeonhardt — APPROVED** [review]
